component.js

```diff
@@ -200,33 +206,35 @@ function factory (options) {
       debug = _shouldComponentUpdate.debug(pattern, logFn);
     }
   }
-
-  function createDefaultArguments (displayName, mixins, render) {
-
-    // (render)
-    if (typeof displayName === 'function') {
-      render      = displayName;
-      mixins      = [];
-      displayName = void 0;
+  
+  var NO_RENDER = "Component definition requires `render` function as its argument"
+
+  // [displayName], [mixins = []], render, [decorate]
+  function createDefaultArguments (/* args */) {
+    var args = Array.prototype.slice.call(arguments);
+    var currArg = 0, argCount = args.length;;
+    
+    var displayName, mixins = [], render, decorate;
+    
+    if (currArg == argCount) throw NO_RENDER;
+    if (typeof args[currArg] === 'string' || args[currArg] instanceof String) {
+      displayName = args[currArg++];
     }
-
-    // (mixins, render)
-    if (typeof displayName === 'object' && typeof mixins === 'function') {
-      render      = mixins;
-      mixins      = displayName;
-      displayName = void 0;
+    
+    if (currArg == argCount) throw NO_RENDER;
+    if (Array.isArray(args[currArg])) {
+      mixins = [args[currArg++]];
+    } else if (typeof args[currArg] === 'object') {
+      mixins = args[currArg++];
     }
-
-    // (displayName, render)
-    if (typeof displayName === 'string' && typeof mixins === 'function') {
-      render = mixins;
-      mixins = [];
+    
+    if (currArg == argCount) throw NO_RENDER;
+    if (typeof args[currArg] === 'function') {
+      render = args[currArg++];
     }
-
-    // Else (displayName, mixins, render)
-
-    if (!Array.isArray(mixins)) {
-      mixins = [mixins];
+    
+    if (currArg != argCount && typeof args[currArg] === 'function') {
```

Comment:
`currArg` is incremented for each successfully found argument. I've thought the check is needed to be sure if we have more arguments. Though I forgot that `typeof undefined == 'undefined'` and further checks will fail anyway.